### PR TITLE
[PT-BR] Type Widening and Narrowing and Type Guards

### PR DESCRIPTION
**Pull request by dheysonalves**
Hi, there was situations that I had to adapt the translations. I don't know if could do that, but couldn't see other way to translate. Hope it's okay to do that. The words widening and narrowing, I kept the english words.

Thanks!

@khaosdoctor, @alvarocamillont, @danilofuchs e @orta

### CODE REVIEW

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
+// Type Guarding é o termo onde você influência o código
+// o fluxo analítico pelo código. TypeScript usa o comportamento existente do javascript
+// no qual valida seus objetos em tempo de execução para influenciar
+// o fluxo do código. Este exemplo assume que você já leu example:code-flow
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Type Guarding é o termo dado para quando você influencia a análise
// de fluxo do código através dele mesmo. O TypeScript usa o comportamento existente do javascript
// o qual valida seus objetos em tempo de execução para influenciar
// o fluxo do código. Este exemplo assume que você já leu example:code-flow
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+// Para rodar através desses exemplos, vamos criar algumas classes,
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Para rodae esses exemplos, vamos criar algumas classes,
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  address: string;
+}
+interface TelephoneOrder extends Order {
+  callerNumber: string;
+}
+interface InternetOrder extends Order {
+  email: string;
+}
+
+// Logo um tipo no qual poderia ser uma dos dois Order subtypes ou undefined
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Aqui um tipo que poderia ser um dos dois subtipos de Order ou undefined
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+type PossibleOrders = TelephoneOrder | InternetOrder | undefined;
+
+// E uma função no qual retorna um PossibleOrder
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// E uma função que retorna um PossibleOrder
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+// Você pode usar o operador Javascript "typeof" para
+// limitar a união. Isso apenas funciona com primitivos
+// dentro do JavaScript (como strings, objects, numbers).
+
+if (typeof possibleOrder === "undefined") {
+  const definitelyNotAnOder = possibleOrder;
+}
+
+// Você pode ver uma lista completa de possíveis typeof values
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Você pode ver uma lista completa de possíveis valores para o typeof
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+// aqui: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/typeof
+
+// Usando operadores JavaScript pode levar você muito longe. Quando
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Usar operadores JavaScript pode levar você muito longe. Quando
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+// você quiser verificar seu próprio tipos de objeto você pode usar
+// type predicate functions.
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// funções com predicado de tipos.
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+// Um type predicate function é uma função onde o tipo de
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Um função com predicado de tipos é uma função onde o tipo de
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+// retorno oferece informação a análise do fluxo de código quando
+// a função retorna verdadeira.
+
+// Usando o pedido possível, nós podemos usar dois type guards
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
// Usando o PossibleOrder, nós podemos usar dois type guards
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+function isATelephoneOrder(order: PossibleOrders): order is TelephoneOrder {
+  return order && "calledNumber" in order;
+}
+
+// Agora nós podemos usar essas funções em condicionais if para estreitar
+// o tipo no qual possibleOrder está dentro do if:
+
+if (isAnInternetOrder(possibleOrder)) {
+  console.log("Order received via email:", possibleOrder.email);
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
  console.log("Pedido recebido por email:", possibleOrder.email);
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+if (isATelephoneOrder(possibleOrder)) {
+  console.log("Order received via phone:", possibleOrder.callerNumber);
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Guards.ts]:
```suggestion
  console.log("Pedido recebido por telefone:", possibleOrder.callerNumber);
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
+// Deve ser mais fácil começar a discussão do
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// Pode ser mais fácil começar a discussão do
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+// widening e narrowing com um exemplo:
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// alargamento e estreitamento com um exemplo:
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+const welcomeString = "Hello There";
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
const welcomeString = "Olá! Estranho";
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+let replyString = "Hey";
+
+// Além das diferenças de texto das strings, welcomeString
+// é um const (no qual significa que o valor nunca mudará)
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// é um const (o que significa que o valor nunca mudará)
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+// e replyString é um let (no qual significa que pode mudar).
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// e replyString é um let (que significa que o valor pode mudar).
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+// Se você passar o mouse sobre as duas variáveis, 
+// obterá informações de tipo muito diferentes do TypeScript:
+//
+//   const welcomeString: "Hello There"
+//
+//   let replyString: string
+
+// TypeScript inferiu que o tipo do welcomeString é
+// a string literal "Hello There", onde replyString
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// a string literal "Hello There", enquanto replyString
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+// é uma string genérica.
+
+// Isso ocorre porque um let precisa ter um tipo mais amplo, você
+// pode determinar replyString para ser qualquer outra string - no qual significa
+// ter uma grande conjunto de possibilidades.
+
+replyString = "Hi :wave:";
+
+// Se replyString tiver uma string literal "Hey" - logo
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// Se replyString tiver uma string literal "Hey" - então
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+// você nunca poderia mudar o valor pois ele poderia mudar
+// para apenas "Hey" de novo.
+
+// Os tipos Widening e Narrowing são sobre expandir e reduzir
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// Os alargamento e estreitamento de tipos são sobre expandir e reduzir
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+// as possibilidades no qual um tipo pode representar.
+
+// Um exemplo do tipo narrowing é trabalhando com uniões, o
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// Um exemplo de estreitamento de tipos é trabalhando com uniões, o
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+// exemplo do fluxo de análise do código é quase inteiramente baseado no
+// narrowing: example:code-flow
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// estreitamento: example:code-flow
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+// O tipo narrowing é o que dá poder ao modo estrito do TypeScript
+// por meio das verificações de nulidade. Com o modo estrito desligado,
+// marcadores de nulidade como undefined e null são ignorados
+// em uma união.
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// O estreitamento de tipos é o que dá poder ao modo estrito do TypeScript
// por meio das verificações de nulidade. Com o modo estrito desligado,
// marcadores de nulidade como undefined e null são ignorados
// em uma união.
```

File: packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+// Normalmente, isso é tão simples quanto uma verificação if:
+
+if (quantumString) {
+  quantumString.length;
+}
+
+// No modo estrito o tipo quantumString tem duas representações.
+// Dentro do if, o tipo foi limitado para apenas string.
+
+// Você pode ver mais exemplos de narrowing aqui:
```
Comment (khaosdoctor) [inline packages/playground-examples/copy/pt/Typescript/Language/Type Widening and Narrowing.ts]:
```suggestion
// Você pode ver mais exemplos de estreitamento aqui:
```

**Review by khaosdoctor — CHANGES_REQUESTED** [review]
Translating some untranslated words

**Comment by dheysonalves** [comment]
Hey @khaosdoctor, have resolved all the proposed changes. If you don't mind, I am a big fan of your dev.to great articles about productivity. It helped me a lot til' today. Thanks a lot.

**Comment by khaosdoctor** [comment]
@Dheyson Hey man! Thanks a lot for the words :D It's always nice to see someone who was helped by one of my contents ❤️

**Review by khaosdoctor — APPROVED** [review]

**Comment by khaosdoctor** [comment]
lgtm

**Pull request merged**